File: src/test/java/dev/nytweetdeck/xapi/live/LivePipelineEventParserTest.java
package dev.nytweetdeck.xapi.live;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import dev.nytweetdeck.xapi.http.XApiHttpException;
import org.junit.jupiter.api.Test;
import tools.jackson.databind.json.JsonMapper;

class LivePipelineEventParserTest {

    private final LivePipelineEventParser parser =
            new LivePipelineEventParser(JsonMapper.builder().build());

    @Test
    void parsesConfirmedWebTweetEngagementEnvelope() {
        var event = parser.parse("""
                {"topic":"/tweet_engagement/123","payload":{"tweet_engagement":{"favorite_count":"8","retweet_count":"3"}}}
                """);

        assertThat(event.type()).isEqualTo("tweet_engagement");
        assertThat(event.entityId()).isEqualTo("123");
        assertThat(event.payload().get("favorite_count").asString()).isEqualTo("8");
    }

    @Test
    void parsesTheSessionConfigurationSentBeforeTopicEvents() {
        var event = parser.parse("""
                {"topic":"/system/config","payload":{"config":{"session_id":"session-1","subscription_ttl_millis":120000}}}
                """);

        assertThat(event.type()).isEqualTo("system_config");
        assertThat(event.entityId()).isNull();
        assertThat(event.payload().get("config").get("session_id").asString())
                .isEqualTo("session-1");
    }

    @Test
    void rejectsUnknownOrMalformedEvents() {
        assertThatThrownBy(() -> parser.parse("""
                {"topic":"/future/1","payload":{"future":{"value":"1"}}}
                """))
                .isInstanceOf(XApiHttpException.class)
                .hasMessageContaining("解析");
    }
}

File: android/app/src/test/java/dev/nytweetdeck/android/xapi/live/LivePipelineEventParserTest.kt
package dev.nytweetdeck.android.xapi.live

import dev.nytweetdeck.android.xapi.XApiException
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Test

class LivePipelineEventParserTest {
    private val parser = LivePipelineEventParser()

    @Test
    fun parsesOfficialSystemConfigAndDefaultsInvalidTtl() {
        val event = parser.parse(
            """{"topic":"/system/config","payload":{"config":{"session_id":"session-1","subscription_ttl_millis":0}}}""",
        ) as ParsedLivePipelineEvent.SystemConfig

        assertEquals("session-1", event.config.sessionId)
        assertEquals(120_000L, event.config.subscriptionTtlMilliseconds)
    }

    @Test
    fun parsesEngagementAndDirectMessageEventsIntoTypedPayloads() {
        val engagement = parser.parse(
            """{"topic":"/tweet_engagement/123","payload":{"tweet_engagement":{"favorite_count":"8","retweet_count":3}}}""",
        ) as ParsedLivePipelineEvent.TweetEngagement
        val update = parser.parse(
            """{"topic":"/dm_update/42","payload":{"dm_update":{"event_id":"e1"}}}""",
        ) as ParsedLivePipelineEvent.DirectMessageUpdate
        val typing = parser.parse(
            """{"topic":"/dm_typing/42","payload":{"dm_typing":{"typing":true}}}""",
        ) as ParsedLivePipelineEvent.DirectMessageTyping

        assertEquals("123", engagement.postId)
        assertEquals(8L, engagement.counts.likeCount)
        assertEquals(3L, engagement.counts.repostCount)
        assertEquals("42", update.userId)
        assertEquals("e1", update.payload.toString().substringAfter("event_id\":\"").substringBefore('"'))
        assertEquals("42", typing.userId)
    }

    @Test
    fun rejectsUnknownOrMalformedEvents() {
        assertThrows(XApiException::class.java) {
            parser.parse("""{"topic":"/future/1","payload":{"future":{"value":"1"}}}""")
        }
        assertThrows(XApiException::class.java) {
            parser.parse("""{"topic":"/system/config","payload":{"config":{"session_id":""}}}""")
        }
    }
}
